### DISCUSSION TITLE
forward and posterior of SingleTask

### DISCUSSION THREAD

**Discussion by skiyohara** [meta-pytorch/botorch#2068] [Q&A] [answered]
Dear botorch users and delopers,

I'm a beginner of botorc hand wondering the difference between forward and posterior method of SingleTask.
To my understanding, the forward method output _N_(0,K), where _N_ is multivariate normal distribution and K is gram matrix, respectively. On the other hand, the posterior method output _N_(0,K+sigma), where sigma is observatoin noise.

However, the predicted standard deviations (pred1.stddev and pred2.stddev) in my case were exacly equal each other.
Here is my toy code.

```
model = SingleTask(X_train,Y_train)
mll = ExactMarginalLogLikelihood(model.likelihood, gp)
fit_gpytorch_model(mll)

pred1 = model(X_train)
pred2 = model.posterior(X_train)

```
Any help would be appreciated.

**Comment by saitcakmak** [comment, marked as answer]
Hi @skiyohara. `Model.forward(...)` evaluates the model prior. It is typically used by GPyTorch while computing the model posterior. What you get when you call `model(X_train)` is `model.__call__(X_train)`, which does a lot more than evaluating `model.forward`. When the model is in `eval` mode, GPyTorch computes the posterior on call. We generally do not recommend directly using these methods with BoTorch models. 

You can think of `Model.posterior` as a wrapper around the GPyTorch posterior computations. In addition to computing the posterior, it applies any necessary transforms to both inputs and outputs, wraps the MVN in a `Posterior` object, and ensures that the output has a consistent shape that is expected by the rest of BoTorch APIs. By default, `Model.posterior` will not add the observation noise to the predictions. It has an `observation_noise` argument that you can set to `True` to include observation noise.

**Reply by skiyohara** [reply]
Thank you for the kind reply. I fully understood.
